backend/internal/oauth/oauth2/dcr/service.go

```diff
@@ -0,0 +1,227 @@
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package dcr
+
+import (
+	"encoding/json"
+	"strings"
+
+	"github.com/asgardeo/thunder/internal/application"
+	"github.com/asgardeo/thunder/internal/application/model"
+	"github.com/asgardeo/thunder/internal/cert"
+	oauth2const "github.com/asgardeo/thunder/internal/oauth/oauth2/constants"
+	oauthutils "github.com/asgardeo/thunder/internal/oauth/oauth2/utils"
+	"github.com/asgardeo/thunder/internal/system/error/serviceerror"
+	"github.com/asgardeo/thunder/internal/system/log"
+)
+
+// DCRServiceInterface defines the interface for the DCR service.
+type DCRServiceInterface interface {
+	RegisterClient(request *DCRRegistrationRequest) (*DCRRegistrationResponse, *serviceerror.ServiceError)
+}
+
+// dcrService is the default implementation of DCRServiceInterface.
+type dcrService struct {
+	appService application.ApplicationServiceInterface
+}
+
+// newDCRService creates a new instance of dcrService.
+func newDCRService(appService application.ApplicationServiceInterface) DCRServiceInterface {
+	return &dcrService{
+		appService: appService,
+	}
+}
+
+// RegisterClient registers a new OAuth client using Dynamic Client Registration.
+func (ds *dcrService) RegisterClient(request *DCRRegistrationRequest) (
+	*DCRRegistrationResponse, *serviceerror.ServiceError) {
+	logger := log.GetLogger().With(log.String(log.LoggerKeyComponentName, "DCRService"))
+
+	if request == nil {
+		return nil, &ErrorInvalidRequestFormat
+	}
+
+	if request.JWKSUri != "" && len(request.JWKS) > 0 {
+		return nil, &ErrorJWKSConfigurationConflict
+	}
+
+	appDTO, err := ds.convertDCRToApplicationDTO(request)
+	if err != nil {
+		logger.Error("Failed to convert DCR request to application DTO", log.String("error", err.Error))
+		return nil, &ErrorServerError
+	}
+
+	createdApp, err := ds.appService.CreateApplication(appDTO)
+	if err != nil {
+		logger.Error("Failed to create application via Application service", log.String("error", err.Error))
+		return nil, ds.mapApplicationErrorToDCRError(err)
+	}
+
+	response, err := ds.convertApplicationToDCRResponse(createdApp, request.ClientName)
+	if err != nil {
+		logger.Error("Failed to convert application to DCR response", log.String("error", err.Error))
+		return nil, err
+	}
+
+	return response, nil
+}
+
+// convertDCRToApplicationDTO converts DCR registration request to Application DTO.
+func (ds *dcrService) convertDCRToApplicationDTO(request *DCRRegistrationRequest) (
```

Comment:
Shall we rename this method? It's not the DTO that's getting passed to the service layer right?